prov/rxd/src/rxd_cq.c

```diff
@@ -229,391 +231,519 @@ static void rxd_ep_recv_data(struct rxd_ep *ep, struct rxd_x_entry *rx_entry,
+static void rxd_handle_cts(struct rxd_ep *ep, struct fi_cq_msg_entry *comp,
+			   struct rxd_pkt_entry *pkt_entry)
 {
-	struct rxd_pkt_entry *pkt_entry, *pkt_entry_head;
-	struct rxd_ctrl_pkt *ctrl_pkt;
-	struct rxd_data_pkt *data_pkt;
-	struct slist_entry *item, *prev;
-	int ret = 0;
+	struct rxd_cts_pkt *cts = (struct rxd_cts_pkt *) (pkt_entry->pkt);
 
-	FI_DBG(&rxd_prov, FI_LOG_EP_CTRL, "got recv completion\n");
+	ep->peers[cts->peer_addr].peer_addr = cts->dg_addr;
 
-	pkt_entry = container_of(comp->op_context, struct rxd_pkt_entry, context);
-	ep->posted_bufs--;
+	rxd_update_peer(ep, cts->peer_addr, cts->dg_addr);
+}
 
-	if (rxd_is_ctrl_pkt(pkt_entry)) {
-		ctrl_pkt = rxd_get_ctrl_pkt(pkt_entry);
-		switch (ctrl_pkt->ctrl_hdr.type) {
-		case RXD_RTS:
-			ret = rxd_handle_rts(ep, comp, pkt_entry);
-			break;
-		case RXD_CTS:
-			rxd_handle_cts(ep, comp, ctrl_pkt);
-			break;
-		case RXD_ACK:
-			rxd_handle_ack(ep, comp, ctrl_pkt);
+static void rxd_handle_ack(struct rxd_ep *ep, struct fi_cq_msg_entry *comp,
+			   struct rxd_pkt_entry *ack_entry)
+{
+	struct rxd_x_entry *tx_entry;
+	struct rxd_ack_pkt *ack = (struct rxd_ack_pkt *) (ack_entry->pkt);
+	struct rxd_pkt_entry *pkt_entry;
+	struct rxd_cq *tx_cq = rxd_ep_tx_cq(ep);
+	struct util_cntr *cntr = ep->util_ep.tx_cntr;
+	fi_addr_t peer = ack->pkt_hdr.peer;
+	struct rxd_pkt_hdr *hdr;
+
+	while (!dlist_empty(&ep->peers[peer].unacked)) {
+		pkt_entry = container_of((&ep->peers[peer].unacked)->next,
+					struct rxd_pkt_entry, d_entry);
+		hdr = rxd_get_pkt_hdr(pkt_entry);
+		if (hdr->seq_no >= ack->pkt_hdr.seq_no)
 			break;
-		default:
-			FI_WARN(&rxd_prov, FI_LOG_EP_CTRL,
-				"Unknown message type\n");
-			goto out;
+
+		if (rxd_pkt_type(pkt_entry) != RXD_DATA) {
+			tx_entry = &ep->tx_fs->entry[hdr->tx_id].buf;
+			if (tx_entry->num_segs > 1) {
+				tx_entry->rx_id = ack->pkt_hdr.rx_id;
+				ep->peers[peer].blocking = 0;
+			}
 		}
-	} else {
-		data_pkt = rxd_get_data_pkt(pkt_entry);
-		rxd_handle_data(ep, comp, data_pkt);
+		dlist_remove(&pkt_entry->d_entry);
+		rxd_release_tx_pkt(ep, pkt_entry);
+	     	ep->peers[peer].unacked_cnt--;
 	}
 
-	pkt_entry_head = container_of(ep->rx_pkt_list.head,
-				      struct rxd_pkt_entry, s_entry);
-	if (pkt_entry_head != pkt_entry) {
-		FI_WARN(&rxd_prov, FI_LOG_EP_CTRL,
-			"matched to incorrect receive\n");
-		slist_foreach(&ep->rx_pkt_list, item, prev) {
-			pkt_entry_head = container_of(item, struct rxd_pkt_entry,
-						      s_entry);
-			if (pkt_entry_head == pkt_entry)
-				break;
+	dlist_foreach_container(&ep->peers[ack->pkt_hdr.peer].tx_list,
+				struct rxd_x_entry, tx_entry, entry) {
+		if (tx_entry->start_seq + (tx_entry->num_segs - 1) >= ack->pkt_hdr.seq_no)
+			break;
+
+		if (!(tx_entry->flags & RXD_NO_COMPLETION)) {
+			fastlock_acquire(&ep->util_ep.tx_cq->cq_lock);
+			tx_cq->write_fn(tx_cq, &tx_entry->cq_entry);
+			fastlock_release(&ep->util_ep.tx_cq->cq_lock);
+			if (cntr)
+				cntr->cntr_fid.ops->add(&cntr->cntr_fid, 1);
```

Comment:
FWIW, we found that calling a no-op function was faster than inserting a if branch into the code.  For a future change, you can define a noop_add() call and drop the if (cntr) check